### ISSUE TITLE
Unit tests from go-lint.yml fail when run locally or with GitHub workflows on a fork

### ISSUE THREAD

**Issue by rluetzner** [minio/minio#14952]
## Expected Behavior

Running unit tests locally, especially using [act](https://github.com/nektos/act) to run the same actions as the GitHub workflows do, should succeed.

## Current Behavior

We noticed that the unit tests do not work on our [MinIO fork](https://github.com/iternity-dotcom/minio). Seeing as there were some changes in the code we went looking for issues, but soon discovered that no matter which commit state we use, the tests always fail.

We've tried the following official release tags (sorted from newest to oldest):

* RELEASE.2022-05-19T18-20-59Z
* RELEASE.2022-05-08T23-50-31Z
* RELEASE.2021-10-13T00-23-17Z (the last tag on which we rebased our fork)
* RELEASE.2021-05-18T00-53-28Z

We've also tried running tests against the current master branch, but they all fail consistently. We made sure that the change from https://github.com/minio/minio/pull/14911 is included, but we're still seeing multiple failures of XLStorageTests, but also other tests.

Here's the log I recorded from running the tests via `act`. The commands to reproduce this are further down below.

[minio-RELEASE.2022-05-19T18-20-59Z.log](https://github.com/minio/minio/files/8738506/minio-RELEASE.2022-05-19T18-20-59Z.log)

## Possible Solution

No idea. I'm immensely confused that the tests seem to have succeeded for https://github.com/minio/minio/pull/14911 to get merged (see https://github.com/minio/minio/actions/runs/2315772676), but don't work when run against that or any other commit locally. Is there something missing from our environment?

## Steps to Reproduce (for bugs)

1. `git clone git@github.com:minio/minio.git && cd minio`
2. `git checkout RELEASE.2022-05-19T18-20-59Z`
3. Make sure Docker is installed.
4. Install act:
   `curl https://raw.githubusercontent.com/nektos/act/master/install.sh | sudo bash`
5. `act -W .github/workflows/go-lint.yml pull_request | tee $HOME/minio-RELEASE.2022-05-19T18-20-59Z.log`
6. `grep 'FAIL:' $HOME/minio-RELEASE.2022-05-19T18-20-59Z.log`

## Context

Not being able to run the existing unit tests locally or on a fork makes it really hard to contribute to your code.

## Regression

I don't know. As mentioned above I tried multiple different commits and even went back pretty far to where `go-lint.yml` was first created. I was hoping to run `git bisect` to figure out when the tests broke, but I could not get them to run anywhere.

## Your Environment

* Operating System and version (`uname -a`): `Linux IT-FR-NB-RS-19 5.10.102.1-microsoft-standard-WSL2 #1 SMP Wed Mar 2 00:30:59 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`

Seeing as the tests are run via Docker, I expect this issue to be independent from the underlying environment.

**Comment by harshavardhana** [comment]
You are running the tests as root user @rluetzner so XLStorage tests will fail since it expects the tests to be run as non root.

**Comment by harshavardhana** [comment]
Also it's your fork - figure it out @rluetzner :-)

**Comment by rluetzner** [comment]
Do I understand correctly that you're using custom GitHub runners that run the workflow containers in a rootless Docker environment? I don't see how that is possible for Windows containers, because AFAIK Docker on Windows does not support rootless mode. The runners also have to be specifically tailored, because GitHub uses does not use rootless Docker environments by default.
Nevertheless I'll try and set something similar up and see if it solves my problem.

**Comment by harshavardhana** [comment]
> Do I understand correctly that you're using custom GitHub runners that run the workflow containers in a rootless Docker environment? I don't see how that is possible for Windows containers, because AFAIK Docker on Windows does not support rootless mode. The runners also have to be specifically tailored, because GitHub uses does not use rootless Docker environments by default. Nevertheless I'll try and set something similar up and see if it solves my problem.

No we do not use custom GitHub runners, we just run on things that are provided by GitHub directly.

If you run as a root user for tests, if will not run. You need to be a regular user.

**Comment by rluetzner** [comment]
Then I don't understand what I'm doing wrong. I'm not running anything as a root user. I've activated the same GitHub workflows for our fork and also tried to run them locally (i.e. using [act](https://github.com/nektos/act)), but they fail. I don't see how I can change the user under which the GitHub workflows are executed.
If you want you can ignore what I'm doing locally. My main point is that I want to get a CI up and running and don't understand why the exact same setup should not work when forked. Is there anything I'm missing? If you're using the regular runners provided by GitHub then everything should just work out of the box, shouldn't it?

**Comment by rluetzner** [comment]
Ah. Sorry for the Docker confusion. I just realized that the GitHub runners are not containers, but VM instances.

Nevertheless I would expect the workflows to succeed out of the box. It's just that my local test setup is not comparable, but the fact that tests are failing on GitHub runners is still baffling me.

**Comment by rluetzner** [comment]
For anybody that might run into a similar problem in the future, we finally figured out what made our tests fail:

1. CVEs will keep getting detected, so they either need to be addressed or ignored in .nancy-ignore.
2. The functional tests clone the latest 'master' of mc and try to build it. In our case that required updating the Go version, because the dependencies in mc's go.mod have since been updated to 1.17.
3. The publically available KES server on play.min.io has been updated since we forked the code. The newer version seems to be incompatible and we had to remove the `KMS` environment variables from the functional tests.

The related pull request can be found here: https://github.com/iternity-dotcom/minio/pull/34 .
